### DISCUSSION TITLE
How to write vertically centered text within a given rect?

### DISCUSSION THREAD

**Discussion by jrom99** [pymupdf/PyMuPDF#1662] [Looking for help] [answered]
Pretty much the question. The code below automatically places both single- and multiline texts just below the top position (as expected), but I don't know how to adjust it in case I want to vertically center the text.

Should I apply a transformation matrix? If so, when? Or should I use the `pos` argument instead? If so, how can I get the required x0 with alignments other than left?

```python
import fitz
doc = fitz.open()
page = doc.new_page()

wrt = fitz.TextWriter(page.rect)
space_available = fitz.Rect(10, 10, page.rect.width-20, 10+page.rect.height/5)
wrt.fill_textbox(space_available, txt, align=1)

page.write_text(writers=wrt)
doc.save("new_doc.pdf")
```

**Comment by JorjMcKie** [comment, marked as answer]
Best have a look at [this](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/textwriter/textwriter-textbox.py) demo script.
It fills a textbox and then writes the TextWriter's to several PDF pages in a variety of ways: with rotations, displacements, other colors or transparency, etc. and also inside some given rectangle.

**Reply by jrom99** [reply]
Thank you very much, I updated the code accordingly and now it does the trick:

```python
import fitz
doc = fitz.open()
page = doc.new_page()

wrt = fitz.TextWriter(page.rect)
space_available = fitz.Rect(10, 10, page.rect.width-20, 10+page.rect.height/5)
wrt.fill_textbox(space_available, txt, align=1)

point = space_available.bl
mat = fitz.Matrix(1,1)
mat.pretranslate(0, -(space_available.height-wrt.text_rect.height)/2)

wrt.write_text(page, morph=(point, mat))

page.write_text(writers=wrt)
doc.save("new_doc.pdf")
```

**Reply by JorjMcKie** [reply]
Exactly! 👍
But this one also works:
[tw-vertical.zip](https://github.com/pymupdf/PyMuPDF/files/8413445/tw-vertical.zip)
